src/storage/gc_worker.rs

```diff
@@ -0,0 +1,279 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::engine::{Engine, ScanMode, StatisticsSummary};
+use super::metrics::*;
+use super::mvcc::{MvccReader, MvccTxn, MAX_TXN_WRITE_SIZE};
+use super::txn::GC_BATCH_SIZE;
+use super::{Callback, Error, Key, Result};
+use kvproto::kvrpcpb::Context;
+use prometheus::local::{LocalHistogram, LocalIntCounter, LocalIntCounterVec};
+use std::fmt::{self, Display, Formatter};
+use std::sync::{Arc, Mutex};
+use util::worker::{self, Builder, Runnable, ScheduleError, Worker};
+
+pub const GC_MAX_PENDING: usize = 2;
+
+struct GCTask {
+    pub ctx: Context,
+    pub safe_point: u64,
+    pub callback: Callback<()>,
+}
+
+impl Display for GCTask {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let epoch = match self.ctx.region_epoch.as_ref() {
+            None => "None".to_string(),
+            Some(e) => format!("{{{}, {}}}", e.conf_ver, e.version),
+        };
+        f.debug_struct("GCTask")
+            .field("region", &self.ctx.region_id)
+            .field("epoch", &epoch)
+            .field("safe_point", &self.safe_point)
+            .finish()
+    }
+}
+
+// `GCRunner` is used to perform GC on the engine
+struct GCRunner {
+    engine: Box<Engine>,
+    ratio_threshold: f64,
+
+    stats: StatisticsSummary,
+    empty_range_counter: LocalIntCounter,
+    skipped_counter: LocalIntCounter,
+    duration_histogram: LocalHistogram,
+    gctask_counter: LocalIntCounter,
+    gctask_fail_counter: LocalIntCounter,
+    gc_keys_counter: LocalIntCounterVec,
+}
+
+impl GCRunner {
+    pub fn new(engine: Box<Engine>, ratio_threshold: f64) -> GCRunner {
+        GCRunner {
+            engine,
+            ratio_threshold,
+
+            stats: StatisticsSummary::default(),
+            empty_range_counter: KV_GC_EMPTY_RANGE_COUNTER.local(),
+            skipped_counter: KV_GC_SKIPPED_COUNTER.local(),
+            duration_histogram: GC_DURATION_HISTOGRAM.local(),
+            gctask_counter: GC_GCTASK_COUNTER.local(),
+            gctask_fail_counter: GC_GCTASK_FAIL_COUNTER.local(),
+            gc_keys_counter: GC_KEYS_COUNTER_VEC.local(),
+        }
+    }
+
+    // Scan keys in the region. Returns scanned keys if any, and a key indicating scan progress
+    fn scan_keys(
+        &mut self,
+        ctx: &Context,
+        safe_point: u64,
+        from: Option<Key>,
+        limit: usize,
+    ) -> Result<(Option<Vec<Key>>, Option<Key>)> {
```

Comment:
Why `Vec<Key>` need a Option wrapper? I think Vec itself can tell the user if it is empty or not.